Build fails using free-threaded Python provided by Conda

I can reproduce this on a Windows machine by running `pip install .` inside a clone of `cryptography` with a conda environment activated that has openssl installed.

To create the environment, do:

```
 micromamba create -n ft-cryptography-build openssl pip python=3.14 python-freethreading
```

Then activate it and set `OPENSSL_DIR`:

```
micromamba activate ft-cryptography-build
$env:OPENSSL_DIR="C:\Users\goldbaum\micromamba\envs\ft-cryptography-build\Library\"
```

And in a clone of the cryptography repo, try to build cryptography from source:

```
python -m pip install .
```

This will fail with:

```
  error: linking with `link.exe` failed: exit code: 1104
    |
    = note: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.44.35207\\bin\\HostX64\\x64\\link.exe" "/DEF:C:\\Users\\goldbaum\\AppData\\Local\\Temp\\rustcv3jlfK\\lib.def" "/NOLOGO" "C:\\Users\\goldbaum\\AppData\\Local\\Temp\\rustcv3jlfK\\symbols.o" "<17 object files omitted>" "C:\\Users\\goldbaum\\cryptography\\target\\release\\deps/{libcryptography_cffi-f48d931899616c5b.rlib,libcryptography_keepalive-8632cf373e0a13ee.rlib,libself_cell-21c2b9426828f958.rlib,libcryptography_x509_verification-c20374e3fc6d46a8.rlib,libcryptography_key_parsing-f8e0b54203f90e6c.rlib,libpem-ab84b43214fa66bc.rlib,libcryptography_openssl-5409ce74ab7d60f9.rlib,libbase64-47c128a8251a9c81.rlib,libcryptography_crypto-2a135def03591f99.rlib,libcryptography_x509-56a26a16a511bc56.rlib,libasn1-af6ae43b922584cc.rlib,libitoa-356ce865e25af8fe.rlib,libpyo3-d1f1edff1852a7f2.rlib,libpyo3_ffi-b5d6e0db9fef7069.rlib,libmemoffset-fc2b7505715378cd.rlib,libunindent-d448b7e1552dd95f.rlib,libopenssl-9dc3077e42eb8719.rlib,libonce_cell-dfaba08cb19d0f40.rlib,libbitflags-035a70ee965cc8a0.rlib,libcfg_if-6684b8fa19753039.rlib,libforeign_types-4f6fe43f096c5a6c.rlib,libforeign_types_shared-6ac562195af208c5.rlib,libopenssl_sys-c2bf494785add1bf.rlib,liblibc-c57205961990c1a9.rlib}.rlib" "<sysroot>\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib/{libstd-*,libpanic_unwind-*,libwindows_targets-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libunwind-*,libcfg_if-*,liballoc-*,librustc_std_workspace_core-*,libcore-*,libcompiler_builtins-*}.rlib" "C:\\Users\\goldbaum\\micromamba\\envs\\ft-cryptography-build\\libs\\python314t.lib" "gdi32.lib" "user32.lib" "crypt32.lib" "ws2_32.lib" "advapi32.lib" "legacy_stdio_definitions.lib" "kernel32.lib" "kernel32.lib" "advapi32.lib" "ntdll.lib" "userenv.lib" "ws2_32.lib" "dbghelp.lib" "/defaultlib:msvcrt" "/NXCOMPAT" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.44.35207\\atlmfc\\lib\\x64" "/LIBPATH:C:\\Users\\goldbaum\\cryptography\\target\\release\\build\\cryptography-cffi-d9eddcf76bfa8f87\\out" "/LIBPATH:C:\\Users\\goldbaum\\micromamba\\envs\\ft-cryptography-build\\Library\\lib" "/LIBPATH:C:\\Users\\goldbaum\\micromamba\\envs\\ft-cryptography-build\\libs" "/OUT:C:\\Users\\goldbaum\\cryptography\\target\\release\\deps\\cryptography_rust.dll" "/OPT:REF,ICF" "/DLL" "/IMPLIB:C:\\Users\\goldbaum\\cryptography\\target\\release\\deps\\cryptography_rust.dll.lib" "/DEBUG" "/PDBALTPATH:%_PDB%" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\libstd.natvis"
    = note: some arguments are omitted. use `--verbose` to show all linker arguments
    = note: LINK : fatal error LNK1104: cannot open file 'python314.lib'â\x90\x8d


  error: could not compile `cryptography-rust` (lib) due to 1 previous error
  ðŸ’¥ maturin failed
    Caused by: Failed to build a native library through cargo
    Caused by: Cargo build finished with "exit code: 101": `"cargo" "rustc" "--profile" "release" "--message-format" "json-render-diagnostics" "--locked" "--manifest-path" "C:\\Users\\goldbaum\\cryptography\\src\\rust\\Cargo.toml" "--lib"`
  Error: command ['maturin', 'pep517', 'build-wheel', '-i', 'C:\\Users\\goldbaum\\micromamba\\envs\\ft-cryptography-build\\python.exe', '--compatibility', 'off'] returned non-zero exit status 1
```

Not at all sure what's going wrong here. Something in the stack is bringing in a requirement for the GIL-enabled libpython314.

This is blocking building cryptography for conda-forge: https://github.com/conda-forge/cryptography-feedstock/pull/163.

That said, as illustrated above, I can reproduce this using the OpenSSL and Python from conda-forge. I'm not sure if it's reproducible using other build toolchains, because I'm not sure how to build openssl from source on my Windows development machine.

Perhaps the use of `uv` in the wheel building workflow is masking some issue that only happens using `pip` and `conda`.

--- a/src/rust/cryptography-cffi/build.rs
+++ b/src/rust/cryptography-cffi/build.rs
@@ -93,6 +93,11 @@ fn main() {
 
     if cfg!(windows) {
         build.define("WIN32_LEAN_AND_MEAN", None);
+        // python.h doesn't set this on the Windows free-threaded build
+        // see https://github.com/python/cpython/issues/127294
+        if is_free_threaded {
+            build.define("Py_GIL_DISABLED", "1");
+        }
     }
 
     build.compile("_openssl.a");

